components/automate-cli/cmd/chef-automate/certRotate.go

```diff
@@ -1151,3 +1214,546 @@ func uniqueIps(ips []string) []string {
+	fontendIps := infra.Outputs.AutomatePrivateIps.Value
+	fontendIps = append(fontendIps, infra.Outputs.ChefServerPrivateIps.Value...)
+	lag, err := getMaxPGLag(log, statusSummary)
+	if err != nil {
+		return err
+	}
+	//////////////////////////////////////////////////////////////////////////
+	if userConsent {
+		agree, err := writer.Confirm(fmt.Sprintf(MAINTENANICE_ON_LAG, lag))
+		if err != nil {
+			return status.Wrap(err, status.InvalidCommandArgsError, errMLSA)
+		}
+		if !agree {
+			return status.New(status.InvalidCommandArgsError, errMLSA)
+		}
+	}
+	err = frontendMaintainenceModeOnOFF(infra, sshConfig, sshUtils, ON, fontendIps, log)
+	if err != nil {
+		return err
+	}
+	//////////////////////////////////////////////////////////////////////////
+
+	waitingStart := time.Now()
+	time.Sleep(waitTime * time.Second)
+	for {
+		lag, err := getMaxPGLag(log, statusSummary)
+		if err != nil {
+			return err
+		}
+		if lag == 0 {
+			break
+		} else {
+			timeElapsed := time.Since(waitingStart)
+			if timeElapsed.Seconds() >= totalWaitTimeOut.Seconds() {
+				return status.Wrap(errors.New(""), status.UnhealthyStatusError, fmt.Sprintf("Follower node is still behind the leader by %d bytes\n", lag))
+			}
+		}
+		time.Sleep(10 * time.Second)
+	}
+	return nil
+}
+
+func getCertsFromTemplate(clusterCertificateFile string) (*CertificateToml, error) {
+	if len(clusterCertificateFile) < 1 {
+		return nil, errors.New("Cluster certificate file is required")
+	}
+	writer.Println("Reading certificates from template file")
+	content, err := fileutils.ReadFile(clusterCertificateFile)
+	if err != nil {
+		writer.Errorln("Error in fetching certificates from template file")
+		return nil, err
+	}
+	certifiacates := &CertificateToml{}
+	toml.Decode(string(content), certifiacates)
+	return certifiacates, nil
+}
+
+func (c *certRotateFlow) certRotateFromTemplate(clusterCertificateFile string, sshUtil SSHUtil, infra *AutomateHAInfraDetails, currentCertsInfo *certShowCertificates, statusSummary StatusSummary, userConsent bool, waitTime time.Duration, totalWaitTimeOut time.Duration) error {
+	sshConfig := c.getSshDetails(infra)
+	configRes := sshutils.SSHConfig{
+		SshUser:    sshConfig.sshUser,
+		SshPort:    sshConfig.sshPort,
+		SshKeyFile: sshConfig.sshKeyFile,
+		HostIP:     sshConfig.hostIP,
+		Timeout:    sshConfig.timeout,
+	}
+	c.log.Debug("==========================================================")
+	c.log.Debug("Stopping traffic MAINTENANICE MODE ON")
+	c.log.Debug("==========================================================")
+	err := checkLagAndStopTraffic(infra, configRes, c.sshUtil, c.log, statusSummary, userConsent, waitTime, totalWaitTimeOut)
+	if err != nil {
+		return err
+	}
+	templateCerts, err := getCertsFromTemplate(clusterCertificateFile)
```

Comment:
we can validate the certificate template before put maintainance mode on

Reply:
we have to put maintenance mode on only in case of PG certificate rotation

Reply:
In case of any error during toml generation for certificate or failed to rotate the certificate, we have to put off the maintenance, other wise, it will create the major downtime

Reply:
we have to switch the maintenance mode on only when, we have proper toml file, all the certificate are good, no typo error from the user in the toml file and we are able to generate the desired toml file.
in case of any error reported during cert rotate we have to switch off the maintenance mode